File: examples/java-examples/src/main/java/com/edgedb/examples/CustomDeserializer.java
package com.edgedb.examples;

import com.edgedb.driver.GelClientPool;
import com.edgedb.driver.annotations.GelDeserializer;
import com.edgedb.driver.annotations.GelName;
import com.edgedb.driver.annotations.GelType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletionStage;

public final class CustomDeserializer implements Example {
    private static final Logger logger = LoggerFactory.getLogger(CustomDeserializer.class);

    @GelType
    public static final class Person {
        private final String name;
        private final Long age;

        @GelDeserializer
        public Person(
                @GelName("name") String name,
                @GelName("age") Long age
        ) {
            this.name = name;
            this.age = age;

            logger.info("Custom deserializer called");
        }
    }

    @Override
    public CompletionStage<Void> run(GelClientPool clientPool) {
        return clientPool.execute("insert Person { name := 'Example', age := 123 } unless conflict on .name")
                .thenCompose(v -> clientPool.queryRequiredSingle(Person.class, "select Person { name, age } filter .name = 'Example'"))
                .thenAccept(person -> logger.info("Person result: {person}"));
    }
}

File: examples/kotlin-examples/src/main/kotlin/com/edgedb/examples/CustomDeserializer.kt
package com.edgedb.examples

import com.edgedb.driver.GelClientPool
import com.edgedb.driver.annotations.GelDeserializer
import com.edgedb.driver.annotations.GelName
import com.edgedb.driver.annotations.GelType
import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory

class CustomDeserializer : Example {
    companion object {
        private val logger = LoggerFactory.getLogger(CustomDeserializer::class.java)!!
    }

    @GelType
    data class Person
    @GelDeserializer
    constructor (
            @GelName("name")
            val name: String,
            @GelName("age")
            val age: Long
    ) {
        init {
            logger.info("Custom deserializer called with: name: {}, age: {}", name, age)
        }
    }

    override suspend fun runAsync(clientPool: GelClientPool) {
        val person = clientPool.queryRequiredSingle(
                Person::class.java,
                """
                    insert Person { name := 'Example', age := 123 } unless conflict on .name;
                    select Person { name, age } filter .name = 'Example'
                """.trimIndent()
        ).await()

        logger.info("Got person: {}", person)
    }
}